### PR TITLE
chore(core): ldml merge from master 2023-02-21 🙀

### PR DESCRIPTION
**Pull request by srl295**
Not 100% happy with this merge. But it gets feature-ldml unbroken.

- core seems to have lost the ability to build deps, so patching something in.
- add new `builder_run_dep_actions()` to the builder for deps
- add dummy test cases to hextobin and keyman-version

Bugs:
- `build.sh configure` would have run npm several times on the deps.  I've changed it to just run `build` on the deps and let them autoconfigure themselves.
- `build.sh clean` does not clean the sub-dependencies at all.  Previously, it seems `build.sh clean` didn't clean any dependencies, so this is probably an improvement.

## Review Note
Please take a look at commit 63be888a1a2d009070596ace38b569ca49846d41 , and even the changes to `core/build.sh` as a whole. Substantial.

#7042

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by srl295** [comment]
https://build.palaso.org/buildConfiguration/Keyman_Common_KPAPI_TestPullRequests_Windows/367453?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=367453_93_93&logView=flowAware

What is it looking for?

Step 1/1: Full meson build and test of Keyman Core (Command Line) 1s
[12:30:22 ](https://build.palaso.org/buildConfiguration/Keyman_Common_KPAPI_TestPullRequests_Windows/367453?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=367453_96_96&logView=flowAware)  Starting: C:\BuildAgent\temp\agentTmp\custom_script7318400013962506378.cmd
[12:30:22 ](https://build.palaso.org/buildConfiguration/Keyman_Common_KPAPI_TestPullRequests_Windows/367453?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=367453_97_97&logView=flowAware)  in directory: C:\BuildAgent\work\99b311828f4ee7c\keyman
[12:30:22 ](https://build.palaso.org/buildConfiguration/Keyman_Common_KPAPI_TestPullRequests_Windows/367453?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=367453_98_98&logView=flowAware)  Check for core/build.sh [Keyman 16+]
[12:30:23 ](https://build.palaso.org/buildConfiguration/Keyman_Common_KPAPI_TestPullRequests_Windows/367453?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=367453_99_99&logView=flowAware)  Invalid parameters. Use --help for help

**Review by mcdurdin — CHANGES_REQUESTED** [review]
I have looked at the last three commits on this PR, which relate to fixing up merge conflicts.

Can we do a bit of a reset on core/build.sh in particular? It looks like you've chosen the master-side of changes and in so doing, lost most of the feature-ldml rewrite of it.

Other comments inline -- once we fixup core/build.sh to be based on the feature-ldml rewrite, then I think most of the other changes in those three commits should be unnecessary.

File: common/web/keyboard-processor/build.sh [hunk 1]
```diff
@@ -40,7 +40,7 @@ if builder_start_action configure; then
 fi
 
 if builder_start_action clean; then
-  npm run clean
+  npm run clean || builder_warn 'npm run clean failed, may not be configured'
```
[the remark this answers is not in the document]
Comment (srl295) [reply common/web/keyboard-processor/build.sh:43]: this particular one, no

File: resources/builder.inc.sh [hunk 2]
```diff
@@ -313,6 +313,8 @@ _builder_run_child_action() {
           if [ -f "$THIS_SCRIPT_PATH/${_builder_target_paths[$target]}/build.sh" ]; then
             _builder_execute_child $action $target
           fi
+        else
+          builder_warn "NO ACTION $action$target"
```
[the remark this answers is not in the document]
Comment (srl295) [reply resources/builder.inc.sh:317]: yes- debugging, thanks.

File: resources/builder.inc.sh [hunk 3]
```diff
@@ -352,6 +354,44 @@ builder_run_child_actions() {
...
+    then
+      # add option flag
+      opts+=("${action}")
+    else
+      _builder_run_dep_action "$action" "${opts[*]}"
+    fi
+    shift
+  done
+}
+
```
[the remark this answers is not in the document]
Comment (srl295) [reply resources/builder.inc.sh:394]: When the other changes go the right way, yes

File: common/web/keyman-version/build.sh [hunk 4]
```diff
@@ -91,6 +92,11 @@ if builder_start_action build; then
   builder_finish_action success build
 fi
 
+if builder_start_action test; then
+  npm test
+  builder_finish_action success test
+fi
+
```
[the remark this answers is not in the document]
Comment (srl295) [reply common/web/keyman-version/build.sh:99]: the goal here was to call the dependencies automatically for each target. Maybe doesn't make sense for test.  I would think we'd want to assume there's a test in each subdir.

**Pull request closed without merging**